pkg/crd/osm/v1/operating_system_profile.go

```diff
@@ -0,0 +1,113 @@
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// OperatingSystemProfileSpec represents the operating system confuration spec.
+	Spec OperatingSystemProfileSpec `json:"spec"`
+}
+
+// OperatingSystemProfileSpec represents the data in the newly created OperatingSystemProfile
+type OperatingSystemProfileSpec struct {
+	// OSType represent the operating system name e.g: ubuntu
+	OSName string `json:"osName"`
+	// OSVersion the version of the operaitng system
+	OSVersion string `json:"osVersion"`
+	// SupportedCloudProviders represent the cloud providers that support the os name and version
+	SupportedCloudProviders []SupportedCloudProvider `json:"supportedCloudProviders"`
+	// Units a list of units that represent the systemd units which will run on the instance
+	Units []Unit `json:"units,omitempty"`
+	// Files a list of files that should exist in the instance
+	Files []File `json:"files,omitempty"`
+}
+
+// SupportedCloudProvider
+type SupportedCloudProvider struct {
+	// Name represents the name of the supported cloud provider
+	Name string `json:"name"`
+	// Spec represents the os/image reference in the supported cloud provider
+	Spec runtime.RawExtension `json:"spec"`
+}
+
+// Unit is a systemd unit used for the operating system config.
+type Unit struct {
+	// Name is the name of a unit.
+	Name string `json:"name"`
+	// Command is the unit's command.
+	Command *string `json:"command,omitempty"`
+	// Enable describes whether the unit is enabled or not.
+	Enable *bool `json:"enable,omitempty"`
+	// Content is the unit's content.
+	Content *string `json:"content,omitempty"`
+	// DropIns is a list of drop-ins for this unit.
+	DropIns []DropIn `json:"dropIns,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
+}
+
+// DropIn is a drop-in configuration for a systemd unit.
+type DropIn struct {
+	// Name is the name of the drop-in.
+	Name string `json:"name"`
+	// Content is the content of the drop-in.
+	Content string `json:"content"`
+}
+
+// File is a file that should get written to the host's file system. The content can either be inlined or
+// referenced from a secret in the same namespace.
+type File struct {
+	// Path is the path of the file system where the file should get written to.
+	Path string `json:"path"`
+	// Permissions describes with which permissions the file should get written to the file system.
+	// Should be defaulted to octal 0644.
+	Permissions *int32 `json:"permissions,omitempty"`
+	// Content describe the file's content.
+	Content FileContent `json:"content"`
+}
+
+// FileContent can either reference a secret or contain inline configuration.
+type FileContent struct {
+	// Inline is a struct that contains information about the inlined data.
+	Inline *FileContentInline `json:"inline,omitempty"`
+}
+
+// FileContentInline contains keys for inlining a file content's data and encoding.
+type FileContentInline struct {
+	// Encoding is the file's encoding (e.g. base64).
+	Encoding string `json:"encoding"`
```

Comment:
We should use an enum for this.

Reply:
Hmm even though I am almost 100% sure that, this field will probably be `base64`